### PR TITLE
bump reth to v1.0.0

### LINKED ISSUE #1249: dev: bump `testcontainers` to `0.19`
**Describe the enhancement request**

At the moment, we are running `testcontainers` version `0.15`:

https://github.com/kkrt-labs/kakarot-rpc/blob/82a84a2d76f56fe42ba3c68cd80580809d56a6ef/Cargo.toml#L128

version `0.19` introduces breaking change when compared to `0.15` and we should migrate to this updated version.

### LINKED ISSUE #1253: feat: bump reth to `1.0` and remove forks
## Description

We need to address several dependency updates to solve [Issue #1193](https://github.com/kkrt-labs/kakarot-rpc/issues/1193) and move forward with our project. The following steps outline the necessary updates and actions required.

### Steps to Complete

1. **Update Dependencies**
   - Update `reth` to `v1.0` and Rust to version `1.79`. This update is in progress in [PR #1245](https://github.com/kkrt-labs/kakarot-rpc/pull/1245).
   - Merge [PR #721](https://github.com/kkrt-labs/ef-tests/pull/721) in the `ef-tests` repository to ensure compatibility with our tests.

2. **Address Rust 1.78 Changes**
   - Rust `1.78` introduced a change regarding unsafe preconditions ([details here](https://blog.rust-lang.org/2024/05/02/Rust-1.78.0.html#asserting-unsafe-preconditions)).
   - This change caused an error in the Dojo repository: [error link](https://github.com/dojoengine/dojo/blob/8f5ee1d525cf22b7db7afc0d1e119e68257a20e5/crates/katana/executor/src/implementation/blockifier/mod.rs#L101).
   - Modify Katana to ensure proper conversion. This fix is in progress in [PR #2129](https://github.com/dojoengine/dojo/pull/2129).

### Action Plan

1. **Merge PR into Katana**
   - Complete and merge the PR in Katana ([PR #2129](https://github.com/dojoengine/dojo/pull/2129)).

2. **Update Dependencies in RPC**
   - Point to the new version of Katana in the RPC and run the CI to verify compatibility.

3. **Merge PR into EF Tests**
   - Complete and merge the PR in EF tests ([PR #721](https://github.com/kkrt-labs/ef-tests/pull/721)).

4. **Update RPC Cargo.toml**
   - Update the provisional pointing in the `rpc` `Cargo.toml` to the latest commit of EF tests.

5. **Restart CI for RPC**
   - Restart the CI for RPC to ensure all changes are properly integrated and tested.

6. **Merge the RPC**
   - After all checks pass, merge the RPC changes.

By following these steps, we can ensure that all dependencies are up-to-date and compatible, enabling us to move forward with the project seamlessly.

### PR DESCRIPTION
**Pull request by tcoratger**
Time spent on this PR:

Should close #1249 
Should close #1253 

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing

# What is the new behavior?

# Does this introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: src/eth_provider/database/mod.rs [hunk 1]
```diff
@@ -79,18 +79,30 @@ impl Database {
         find_options: impl Into<Option<FindOpts>>,
     ) -> DatabaseResult<Vec<T>>
     where
-        T: DeserializeOwned + CollectionName,
+        T: DeserializeOwned + CollectionName + Sync + Send,
```
[the remark this answers is not in the document]
Comment (tcoratger) [reply src/eth_provider/database/mod.rs:82]: yes after Mongo 3.0 this is needed for the collection: https://github.com/mongodb/mongo-rust-driver/blob/b781af26dfb17fe62823a866a025de9fb102e0b3/src/db.rs#L120

File: src/eth_provider/database/mod.rs [hunk 2]
```diff
@@ -67,7 +67,7 @@ impl Database {
     /// Returns a collection from the database.
     pub fn collection<T>(&self) -> Collection<T>
     where
-        T: CollectionName,
+        T: CollectionName + Sync + Send,
```
[the remark this answers is not in the document]
Comment (tcoratger) [reply src/eth_provider/database/mod.rs:70]: yes after Mongo 3.0 this is needed for the collection: https://github.com/mongodb/mongo-rust-driver/blob/b781af26dfb17fe62823a866a025de9fb102e0b3/src/db.rs#L120

File: src/test_utils/mongo/mod.rs [hunk 3]
```diff
@@ -68,13 +85,13 @@ pub enum CollectionDB {
...
+    /// Represents a stored header associated with a `CollectionDB`.
     StoredHeader(StoredHeader),
-    /// Represents a stored transaction associated with a CollectionDB.
+    /// Represents a stored transaction associated with a `CollectionDB`.
     StoredTransaction(StoredTransaction),
-    /// Represents a stored transaction receipt associated with a CollectionDB.
+    /// Represents a stored transaction receipt associated with a `CollectionDB`.
     StoredTransactionReceipt(StoredTransactionReceipt),
-    /// Represents a stored log associated with a CollectionDB.
+    /// Represents a stored log associated with a `CollectionDB`.
```
[the remark this answers is not in the document]
Comment (tcoratger) [reply src/test_utils/mongo/mod.rs]: fixed

File: src/eth_provider/database/mod.rs [hunk 4]
```diff
@@ -79,18 +79,30 @@ impl Database {
...
     {
         let find_options = find_options.into();
-        Ok(self.collection::<T>().find(filter, find_options.unwrap_or_default().build()).await?.try_collect().await?)
+        Ok(self
+            .collection::<T>()
+            .find(Into::<Option<Document>>::into(filter).unwrap_or_default())
+            .with_options(find_options.unwrap_or_default().build())
+            .await?
+            .try_collect()
+            .await?)
```
[the remark this answers is not in the document]
Comment (tcoratger) [reply src/eth_provider/database/mod.rs:91]: Agreed, fixed

File: src/eth_provider/database/mod.rs [hunk 5]
```diff
@@ -130,41 +142,41 @@ impl Database {
     where
         T: DeserializeOwned + Unpin + Send + Sync + CollectionName,
     {
-        let find_one_option = FindOneOptions::builder().sort(sort).build();
-        Ok(self.collection::<T>().find_one(filter, find_one_option).await?)
+        Ok(self
```
[the remark this answers is not in the document]
Comment (tcoratger) [reply src/eth_provider/database/mod.rs:142]: Agreed, fixed

File: Cargo.toml [hunk 6]
```diff
@@ -22,20 +22,20 @@ rust.missing_debug_implementations = "warn"
...
 # new lints that are still under development
-nursery = "warn"
+nursery = { level = "warn", priority = -1 }
 # avoid lints that are too pedantic
 future_not_send = "allow"
 fallible_impl_from = "allow"
 
 # lints which are rather strict or have occasional false positives
-pedantic = "warn"
+pedantic = { level = "warn", priority = -1 }
```
Comment (greged93) [inline Cargo.toml:38]: Is there a reason to set the priority of these lints higher?

File: src/test_utils/mongo/mod.rs [hunk 7]
```diff
@@ -52,6 +52,23 @@ pub fn generate_port_number() -> u16 {
     local_addr.port()
 }
 
+#[derive(Default, Debug)]
+pub struct MongoContainer;
```
Comment (greged93) [inline src/test_utils/mongo/mod.rs:56]: Maybe MongoImage?

File: Cargo.toml (hunk 6 above)
Comment (tcoratger) [reply Cargo.toml:38]: It is something nice that was detected by the clippy on the latest version of Rust:

https://rust-lang.github.io/rust-clippy/master/#/lint_groups_priority

As we want to allow certain clippy, for overrides properly, we must set the clippy warning of the same category above with a lower priority. Quite well described in the error:

```shell
error: lint group `pedantic` has the same priority (0) as a lint
 --> Cargo.toml:38:1
 |
38 | pedantic = "warn"
 | ^^^^^^^^ ------ has an implicit priority of 0
39 | # avoid lints that are too pedantic
40 | must_use_candidate = "allow"
 | ------------------ has the same priority as this lint
 |
 = note: the order of the lints in the table is ignored by Cargo
 = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#lint_groups_priority
 = note: `-D clippy::lint-groups-priority` implied by `-D warnings`
 = help: to override `-D warnings` add `#[allow(clippy::lint_groups_priority)]`
help: to have lints override the group set `pedantic` to a lower priority
 |
38 | pedantic = { level = "warn", priority = -1 }
 | ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

error: could not compile `kakarot-rpc` (lib) due to 1 previous error
```

File: src/test_utils/mongo/mod.rs (hunk 7 above)
Comment (tcoratger) [reply src/test_utils/mongo/mod.rs]: yes agreed, this is better, fixed

**Review by greged93 — APPROVED** [review]

**Pull request merged**